compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -833,6 +835,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         typed(desugar.makeCaseLambda(tree.cases, protoFormals.length, unchecked) withPos tree.pos, pt)
       case _ =>
         val sel1 = typedExpr(tree.selector)
+        if (sel1.tpe.isPhantom)
+          ctx.error(MatchOnPhantom(), sel1.pos)
         val selType = widenForMatchSelector(
```

Comment:
Currently I do not see any use case for matching on a phantom values as this value has no representation at runtime. We would already know it's most specific type.